### ISSUE TITLE
CP-SAT: Performance drops from 60s to 700s - why?

### ISSUE THREAD

**Issue by adihe** [google/or-tools#1102]
Hello all

First of all, thanks to the hard working team for providing CP-SAT as open source project. It looks very promising.

I am investigating whether CP-SAT could serve as solver component for our nurse shift scheduling solution. I used your following shift scheduling example as a basis for a prototype:
https://github.com/google/or-tools/blob/master/examples/dotnet/ShiftSchedulingSat.cs

CP-SAT looks very promising so far but I am running into performance problems with the prototype we built.

The problem:
Adding an additional type of constraint ("cover constraint") caused a drop in solving performance from roughly 1 minute to roughly 12 minutes. I am not an expert in constraint programming but I do have the impression that a drop by 12x is way too high.

The additional constraint:
```cs
            // Cover constraints
            foreach (int s in Range(numShifts))
            {
                foreach (int w in Range(numWeeks))
                {
                    foreach (int d in Range(7))
                    {
                        var works = new List<IntVar>();
                        foreach (int e in Range(numEmployees))
                        {
                            if(work[e, s, w * 7 + d] != null)
                                works.Add(work[e, s, w * 7 + d]);
                        }
                        IntVar[] wa = works.ToArray();

                        var optDemand = weeklyCoverDemands[d, s];
                        optDemand = Math.Min(numEmployees, optDemand);

                        if (optDemand > 0)
                            model.Add(wa.Sum() <= optDemand); // don't staff more people than optimal-demand
                    }
                }
            }

```
A longer code snippet from the prototype can be seen here:
https://pastebin.com/ssWUTH8F

My question:
* Is this performance drop to be expected - is it by design of CP-SAT?
* Can I improve/fix my code to reduce the performance drop?

Thank you very much for your insights!

Best regards,
Adrian

**Comment by lperron** [comment]
You are adding a lot of constraints, no? And you are making the problem much harder.
The model already penalyze over-capacity.
Why don't you just try to increase the over-capacity penalty?

**Comment by adihe** [comment]
Hello Laurent

No there is no other excess penalty. 

Yes, my model contains roughly 80 nurses and 30 shifts so it ends up adding more constraints than the original example.

Compared to a real life case however the constraints are very few. A real life case will easily contain additional 15 different types of constraints.

I am currently using a TabuSearch algorithm that yields "ok" results in 2 minutes.

Are there any tools available that would help me troubleshoot and profile cp-sat performance in my prototype?

Thanks,
Adrian

**Comment by lperron** [comment]
There is:

https://github.com/google/or-tools/blob/master/examples/dotnet/ShiftSchedulingSat.cs#L117

currently the model allow excess shifts, but penalizes them.
Your model forbids any excess.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by lperron** [comment]
About testing the model, not much.
You can have a look at number of branches per seconds to estimate how heavy
the model is.

Can you try adding parallelism

solver.StringParameters = "numSearchWorkers:8";
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by stuarthillary** [comment]
Note the string parameters to enable parallelism in C# should read

solver.StringParameters = "num_search_workers:8";

**Comment by adihe** [comment]
Thank you very much for your suggestion, I will try this and get back to you. 

BR, 
Adrian

**Comment by adihe** [comment]
The configuration change "num_search_workers:8" did not help, the performance dropped further.

I rewrote the prototype in large parts and the problem has gone since. Unfortunately I cannot tell exactly why it has gone.

So far I came up with the following rules of thumb regarding designing cp-sat models for good performance, some of them of course rather obvious:
- Keep the amount of variables and constraints as low as possible
- Keep value range of each variable as narrow as possible and prefer boolean variables over int variables
- Constrain the sum of variables (e.g. works.sum() ==0) instead of constraining each individual variable (e.g. work[e,s,d] == 0)
- Replace constraints (e.g. model.Add()) by optimization function (e.g. model.Minimize(xyz)) wherever possible
- Prefer approaches with a AddBoolOr constraint over approaches that use constraints with OnlyEnforceIf
- Precalculate your model to make sure that you don't create contraints that are impossible to hit (I found that CP-SAT does not do this automatically yet)
- Segregate problem into smaller problems and merge the results - if possible
- Structure your problem into steps that are easy to turn on/off in order to troubleshoot constraints that cause performance impacts

Any comments on this?
Do you have any further general performance recommendations?

**Comment by lperron** [comment]
Good summary of how to write a good model :-)

A few comments:

   - Prefer approaches with a AddBoolOr constraint over approaches that use
   constraints with OnlyEnforceIf

I would rephrase that as: prefer boolean constraints over other integer
constraints.

I do not understand:

   - Precalculate your model to make sure that you don't create contraints
   that are impossible to hit (I found that CP-SAT does not do this
   automatically yet)


About recommendations, modeling is an art and we have very little
background on modeling with CP-SAT.
Good job.


Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by adihe** [comment]
Hi Laurent

Thanks! 

Let me explain what I meant by that bullet. Imagine the nurse scheduling problem. You have a model and you feed data (nurses,shifts,fixed assignments,etc) into it. If you analyze the data you might figure out that some nurses do have a day off each Friday/Saturday/Sunday and therefore these particular nurses do not need constraints such as "shift X not longer than 4 days in a row". So you can reduce the number of constraints by preprocessing/analyzing the data.

**Comment by lperron** [comment]
Yes, but just solving wise, I am not sure it is worth the effort :-)

But, doing analysis on all the solutions are always worth it.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Comment by adihe** [comment]
> Replace constraints (e.g. model.Add()) by optimization function (e.g. model.Minimize(xyz)) wherever possible

Update on this:
If the constraint significantly reduces the solution space it may be worth implementing it as a hard constraint (such as model.Add(..)) instead of optimization function (e.g. model.Minimize(xyz)). I am using a technique where I solve the model first with a hard constraint approach and if I get a an infeasible solution I perform a fallback and solve again with the optimization function instead. You could call this technique "hard-constraint-relaxation". It improves overall performance in case you have many sub-models where only few of them will require a fallback retry.

**Comment by lperron** [comment]
Hi, 

Is this still active ? 
Have you tried with 7.2 ?

Thanks

**Comment by adihe** [comment]
It is resolved. Thank you.

**Comment by zarglayoun** [comment]
> * Keep the amount of variables and constraints as low as possible

In thoery, in constraint programming adding new constraints leads to faster solutions. What's your take on this in practice using or-tools ?

**Comment by lperron** [comment]
This is not always true. Additional constraints can help if they propagate,
can slow down if they are expensive and ineffective.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le mar. 10 nov. 2020 à 19:53, zarglayoun <notifications@github.com> a
écrit :

**Comment by zarglayoun** [comment]
Do you have an example when adding aome constraints the problem take more time to be solved (assuming that the number of variables does not change).

**Comment by zarglayoun** [comment]
hh
